File: machine/cpp/machine/devices/system/debug_transmit.cpp
#include "machine/devices/system/debug_transmit.h"

#include "common/utf8.h"
#include "machine/memory/memory.h"

namespace bmsx {

SystemDebugTransmit::SystemDebugTransmit(Memory& memory)
	: m_memory(memory) {
	memory.mapIoWrite<&SystemDebugTransmit::writeChar>(IO_SYS_PRINT_CHAR, *this);
	memory.mapIoWrite<&SystemDebugTransmit::flushLine>(IO_SYS_PRINT_FLUSH, *this);
}

void SystemDebugTransmit::reset() {
	clearOutput();
	m_charWord = 0u;
	m_flushWord = 0u;
	m_memory.writeIoU32(IO_SYS_PRINT_CHAR, 0u);
	m_memory.writeIoU32(IO_SYS_PRINT_FLUSH, 0u);
}

auto SystemDebugTransmit::captureState() const -> SystemDebugTransmitState {
	return {
		m_charWord,
		m_flushWord,
	};
}

void SystemDebugTransmit::restoreState(const SystemDebugTransmitState& state) {
	clearOutput();
	m_charWord = state.charWord;
	m_flushWord = state.flushWord;
	m_memory.writeIoU32(IO_SYS_PRINT_CHAR, state.charWord);
	m_memory.writeIoU32(IO_SYS_PRINT_FLUSH, state.flushWord);
}

auto SystemDebugTransmit::availableByteCount() const -> u32 {
	return m_completeByteCount;
}

auto SystemDebugTransmit::readByte() -> u8 {
	const u8 value = m_outputBytes[m_outputReadIndex];
	m_outputReadIndex = (m_outputReadIndex + 1u) & (SYS_PRINT_BUFFER_BYTES - 1u);
	m_outputByteCount -= 1u;
	m_completeByteCount -= 1u;
	return value;
}

void SystemDebugTransmit::writeChar([[maybe_unused]] u32 address, u32 value) {
	m_charWord = value;
	const u32 byteCount = encodeUtf8Codepoint(value, m_encodingBytes);
	if (!reserveBytes(byteCount)) {
		return;
	}
	for (u32 index = 0u; index < byteCount; ++index) {
		appendByte(m_encodingBytes[index]);
	}
}

void SystemDebugTransmit::flushLine([[maybe_unused]] u32 address, u32 value) {
	m_flushWord = value;
	if (reserveBytes(1u)) {
		appendByte(static_cast<u8>('\n'));
		m_completeByteCount = m_outputByteCount;
	}
	m_lineOverflowed = false;
}

auto SystemDebugTransmit::reserveBytes(u32 byteCount) -> bool {
	if (m_lineOverflowed) {
		return false;
	}
	if (m_outputByteCount + byteCount <= SYS_PRINT_BUFFER_BYTES) {
		return true;
	}
	m_outputByteCount = m_completeByteCount;
	m_lineOverflowed = true;
	return false;
}

void SystemDebugTransmit::clearOutput() {
	m_outputReadIndex = 0u;
	m_outputByteCount = 0u;
	m_completeByteCount = 0u;
	m_lineOverflowed = false;
}

void SystemDebugTransmit::appendByte(u8 value) {
	m_outputBytes[(m_outputReadIndex + m_outputByteCount) & (SYS_PRINT_BUFFER_BYTES - 1u)] = value;
	m_outputByteCount += 1u;
}

} // namespace bmsx

File: machine/ts/machine/devices/system/debug_transmit.ts
import { encodeUtf8Codepoint } from '../../../common/utf8';
import {
	IO_SYS_PRINT_CHAR,
	IO_SYS_PRINT_FLUSH,
	SYS_PRINT_BUFFER_BYTES,
} from '../../../spec/bmsx/io';
import { Memory } from '../../memory/memory';

export type SystemDebugTransmitState = {
	charWord: number;
	flushWord: number;
};

const ASCII_NEWLINE = 10;

export class SystemDebugTransmit {
	private readonly outputBytes = new Uint8Array(SYS_PRINT_BUFFER_BYTES);
	private outputReadIndex = 0;
	private outputByteCount = 0;
	private completeByteCount = 0;
	private lineOverflowed = false;
	private readonly encodingBytes = new Uint8Array(4);

	public constructor(private readonly memory: Memory) {
		memory.mapIoWrite(IO_SYS_PRINT_CHAR, this, SystemDebugTransmit.writeChar);
		memory.mapIoWrite(IO_SYS_PRINT_FLUSH, this, SystemDebugTransmit.flushLine);
	}

	public reset(): void {
		this.clearOutput();
		this.memory.writeIoU32(IO_SYS_PRINT_CHAR, 0);
		this.memory.writeIoU32(IO_SYS_PRINT_FLUSH, 0);
	}

	public captureState(): SystemDebugTransmitState {
		return {
			charWord: this.memory.readIoU32(IO_SYS_PRINT_CHAR),
			flushWord: this.memory.readIoU32(IO_SYS_PRINT_FLUSH),
		};
	}

	public restoreState(state: SystemDebugTransmitState): void {
		this.clearOutput();
		this.memory.writeIoU32(IO_SYS_PRINT_CHAR, state.charWord);
		this.memory.writeIoU32(IO_SYS_PRINT_FLUSH, state.flushWord);
	}

	public availableByteCount(): number {
		return this.completeByteCount;
	}

	public readByte(): number {
		const value = this.outputBytes[this.outputReadIndex];
		this.outputReadIndex = (this.outputReadIndex + 1) & (SYS_PRINT_BUFFER_BYTES - 1);
		this.outputByteCount -= 1;
		this.completeByteCount -= 1;
		return value;
	}

	private static writeChar(context: SystemDebugTransmit, _address: number, value: number): void {
		const byteCount = encodeUtf8Codepoint(value, context.encodingBytes);
		if (!context.reserveBytes(byteCount)) {
			return;
		}
		for (let index = 0; index < byteCount; index += 1) {
			context.appendByte(context.encodingBytes[index]);
		}
	}

	private static flushLine(context: SystemDebugTransmit): void {
		if (context.reserveBytes(1)) {
			context.appendByte(ASCII_NEWLINE);
			context.completeByteCount = context.outputByteCount;
		}
		context.lineOverflowed = false;
	}

	private reserveBytes(byteCount: number): boolean {
		if (this.lineOverflowed) {
			return false;
		}
		if (this.outputByteCount + byteCount <= SYS_PRINT_BUFFER_BYTES) {
			return true;
		}
		this.outputByteCount = this.completeByteCount;
		this.lineOverflowed = true;
		return false;
	}

	private clearOutput(): void {
		this.outputReadIndex = 0;
		this.outputByteCount = 0;
		this.completeByteCount = 0;
		this.lineOverflowed = false;
	}

	private appendByte(value: number): void {
		this.outputBytes[(this.outputReadIndex + this.outputByteCount) & (SYS_PRINT_BUFFER_BYTES - 1)] = value;
		this.outputByteCount += 1;
	}
}
